internal/client/oidc/client.go

```diff
@@ -60,105 +61,212 @@ var errRefreshAccessToken = fmt.Errorf("Failed refreshing access token")
 
 var oidcScopes = []string{oidc.ScopeOpenID, oidc.ScopeOfflineAccess, oidc.ScopeEmail}
 
-// OidcClient is a structure encapsulating an HTTP client, OIDC transport, and a token for OpenID Connect (OIDC) operations.
-type OidcClient struct {
+// OIDCClient is a structure encapsulating an HTTP client, OIDC transport, and OIDC context (token, trust tupple) for OpenID Connect (OIDC) operations.
+type OIDCClient struct {
 	httpClient    *http.Client
 	oidcTransport *oidcTransport
-	tokens        *oidc.Tokens[*oidc.IDTokenClaims]
-	tokensFile    string
+
+	oidcContextMu   sync.Mutex
+	oidcContext     OIDCContext
+	oidcContextFile string
+
+	authenticateOpenBrowser bool
+	authenticateCallback    func(tokenURL string)
+}
+
+// OIDCContext holds the OIDC context information, which is required to
+// authenticate with OIDC and to refresh the OIDC tokens if necessary.
+type OIDCContext struct {
+	TrustTuple OIDCTrustTuple                   `json:"trust_tuple"`
+	Tokens     oidc.Tokens[*oidc.IDTokenClaims] `json:"tokens"`
 }
 
-// NewClient constructs a new OidcClient, ensuring the token field is non-nil to prevent panics during authentication.
-func NewClient(httpClient *http.Client, tokensFile string) *OidcClient {
-	client := OidcClient{
-		tokens:        loadTokensFromFile(tokensFile),
-		tokensFile:    tokensFile,
-		httpClient:    httpClient,
-		oidcTransport: &oidcTransport{},
+// OIDCTrustTuple is the issuer, clientid and the audience shared by the server
+// in order to authenticate (or refresh) with the OIDC relying party.
+type OIDCTrustTuple struct {
+	Issuer   string `json:"issuer"`
+	ClientID string `json:"client_id"`
+	Audience string `json:"audience"`
+}
+
+type ClientOption func(c *OIDCClient)
+
+// WithoutOpenBrowser disables the automated opening of the web browser to perform
+// the user's authentication flow with the OIDC rely party. In this case
+// the client will just print the URL and the code to the console.
+//
+// This is helpful in headless scenarios or in tests.
+func WithoutOpenBrowser() ClientOption {
+	return func(c *OIDCClient) {
+		c.authenticateOpenBrowser = false
 	}
+}
+
+// WithAuthenticateCallback takes a callback function, which is called after
+// the authentication flow has been started and the token has been received.
+// The URL to confirm the token is passed by to the callback function and
+// can be queried in the callback allowing for automation.
+//
+// This is mainly useful in tests, e.g. with mini-oidc, which does not do any
+// authentication and just issues the tokens when the URL is queried.
+func WithAuthenticateCallback(callback func(tokenURL string)) ClientOption {
+	return func(c *OIDCClient) {
+		c.authenticateCallback = callback
+	}
+}
```

Comment:
Since we only use these two functions in tests (and probably, they should not be used in production code), we could also move them to a file `client_internal_test.go`, making them only available in tests.

@stgraber what do you think?